### DISCUSSION TITLE
WARNING:apscheduler.scheduler:Error getting due jobs from job store 'sqlalchemy': columns argument to select() must be a Python list or other iterable

### DISCUSSION THREAD

**Discussion by AleksSpace** [agronholm/apscheduler#733] [Q&A] [answered]
I am trying to configure the scheduler so that it saves tasks to the DB.

`jobstores = {
    'sqlalchemy': SQLAlchemyJobStore(url=SQLALCHEMY_DATABASE_URI),
}
executors = {
    'default': {'type': 'threadpool', 'max_workers': 20},
    'processpool': ProcessPoolExecutor(max_workers=10)
}
job_defaults = {
    'coalesce': True,
    'max_instances': 30
}

scheduler = BackgroundScheduler()
scheduler.configure(
    jobstores=jobstores,
    executors=executors,
    job_defaults=job_defaults
)
`
but when I start, I get an error:
WARNING:apscheduler.scheduler:Error getting due jobs from job store 'sqlalchemy': columns argument to select() must be a Python list or other iterable.
please help me make everything work!

**Comment by agronholm** [comment, marked as answer]
APScheduler 3.x has not been adapted for use with SQLAlchemy 2. You should be able to get it working with 1.4.x. APScheduler 4 (still in alpha) does work with SQLAlchemy 2.

**Reply by AleksSpace** [reply]
Thank you)) how do I install APScheduler 4? will it work with python 3.6?

**Reply by agronholm** [reply]
APScheduler 4 does not work with Python 3.6. The PSF dropped support for Python 3.6 1.5 years ago so you shouldn't be using it anymore either. You can install APScheduler 4 with `pip install --pre -U APScheduler`. If you want to try the latest version in master, you can do `pip install git+https://github.com/agronholm/apscheduler.git`.

**Reply by AleksSpace** [reply]
thank you very much)))

**Reply by agronholm** [reply]
The least painful way for you, if you insist on sticking with Python 3.6, is to downgrade SQLAlchemy to v1.4.x.

**Reply by AleksSpace** [reply]
I have an old project, the Flask-SQLAlchemy==2.1. When I upgrade it to 1.4.x, I get an error:
ERROR:flask_appbuilder.security.sqla.manager:DB Creation and initialization failed: get_bind() got an unexpected keyword argument 'bind'
could you tell me what to do, please?

**Reply by AleksSpace** [reply]
a Flask-based application

**Reply by agronholm** [reply]
You could go farther back, to sqlalchemy 1.3. That at least should work.

**Reply by agronholm** [reply]
Why are the dependencies in that application not pinned? This is the exact kind of problem that occurs when not pinning dependencies to known good versions.

**Reply by AleksSpace** [reply]
sorry, I made a mistake, I have SQLAlchemy versions==1.3.20. and the error is still the same

**Reply by agronholm** [reply]
So which version of APScheduler did you try this with? I checked the latest 3.x release and I did in fact adapt it to use SQLAlchemy 1.4.x; it's even required in the dependencies for the `sqlalchemy` extra.

**Comment by AleksSpace** [comment]
I tried to install:
pip install git+https://github.com/agronholm/apscheduler.git
But there was an error:
ERROR: Command errored out with exit status 1:
   command: 'd:\dev\job\test_pandas\venv\scripts\python.exe' 'd:\dev\job\test_pandas\venv\lib\site-packages\pip' install --ignore-installed --no-user --prefix 'C:\Users\Zaiki\AppData\Local\Temp\pip-build-env-jz15xfe1\overlay' --no-warn-script-location --no-binary :none: --only-binary :none: 
-i https://pypi.org/simple -- 'setuptools >= 64' 'setuptools_scm >= 6.4'

**Reply by agronholm** [reply]
Was there no error message whatsoever?

**Reply by AleksSpace** [reply]
Sorry, I didn't see your message.

**Comment by AleksSpace** [comment]
Please tell me, does APScheduler 4 work stably with python version 3.11?

**Reply by agronholm** [reply]
It's still in alpha, so stability is not even implied. That said, it was tested from the start against Python 3.11. The only issues I can think of that could arise in some rare situations relate to AnyIO and task uncancellation which is a specific quirk of Python 3.11.

**Reply by AleksSpace** [reply]
So now the most stable version is 3.10.1? And it works well with python 3.11?

**Reply by agronholm** [reply]
I can only say that it gets automatically tested against Python 3.11 and passes. I haven't actually used it myself in a long time.
